main/src/ca/uwaterloo/flix/language/errors/SafetyError.scala

```diff
@@ -137,14 +137,93 @@ object SafetyError {
          |
          |${code(loc, "the upcast occurs here.")}
          |
-         |Actual type:      ${actual.tpe}
-         |Tried casting to: ${expected.tpe}
+         |Actual type:      $actual
+         |Tried casting to: $expected
+         |""".stripMargin
+    }
+
+    override def explain(formatter: Formatter): Option[String] = None
+  }
+
+  /**
+    * An error raised to indicate an invalid use of supercast.
+    *
+    * @param actual   the type of the expression being supercast.
+    * @param expected the type being cast to, i.e. the type of the supercast expression itself.
+    * @param loc      the source location of the supercast.
+    */
+  case class UnsafeSupercast(actual: Type, expected: Type, loc: SourceLocation) extends SafetyError {
+    override def summary: String = "Unsafe supercast."
+
+    override def message(formatter: Formatter): String = {
+      import formatter._
+      s"""${line(kind, source.name)}
+         |>> The following supercast is unsafe and not allowed.
+         |
+         |${code(loc, "the supercast occurs here.")}
+         |
+         |Actual type:      $actual
+         |Tried casting to: $expected
          |""".stripMargin
     }
 
     override def explain(formatter: Formatter): Option[String] = None
   }
 
+  /**
+    * An error raised to indicate use of supercast on a non java type.
+    *
+    * @param nonJavaType the type that is **not** a java type.
+    * @param other       the other type.
+    * @param loc         the source location of the supercast.
+    */
+  case class NonJavaTypeSupercast(nonJavaType: Type, other: Type, loc: SourceLocation) extends SafetyError {
+    override def summary: String = "Unsafe supercast"
+
+    override def message(formatter: Formatter): String = {
+      import formatter._
+      s"""${line(kind, source.name)}
+         |>> The following supercast tries to cast a non-java to a java type.
```

Comment:
Capitalize Java :)

Reply:
I'm not sure "Java type" is the right verbiage. But I cannot think of something better ATM.